rust/core-lib/src/editor.rs

```diff
@@ -13,166 +13,110 @@
-const MAX_SIZE_LIMIT: usize = 1024 * 1024;
-
-enum CharacterEncoding {
-    Utf8,
-    Utf8WithBom
-}
-
-const UTF8_BOM: &str = "\u{feff}";
-
 enum IndentDirection {
     In,
     Out
 }
 
+//TODO: move me
 fn last_selection_region(regions: &[SelRegion]) -> Option<&SelRegion> {
     for region in regions.iter().rev() {
         if !region.is_caret() {
             return Some(region);
         }
     }
-
     None
 }
 
 pub struct Editor {
+    /// The contents of the buffer.
     text: Rope,
-    encoding: CharacterEncoding,
-
-    path: Option<PathBuf>,
-    file_mod_time: Option<SystemTime>,
-    file_has_changed: bool,
-    buffer_id: BufferIdentifier,
-    syntax: SyntaxDefinition,
-    view: View,
+    /// The CRDT engine, which tracks edit history and manages concurrent edits.
     engine: Engine,
+
+    /// The most recent revision.
     last_rev_id: RevId,
+    /// The revision of the last save.
     pristine_rev_id: RevId,
     undo_group_id: usize,
-    live_undos: Vec<usize>, //  undo groups that may still be toggled
-    cur_undo: usize, // index to live_undos, ones after this are undone
-    undos: BTreeSet<usize>, // undo groups that are undone
-    gc_undos: BTreeSet<usize>, // undo groups that are no longer live and should be gc'ed
+    /// Undo groups that may still be toggled
+    live_undos: Vec<usize>,
+    /// The index of the current undo; subsequent undos are currently 'undone'
+    /// (but may be redone)
+    cur_undo: usize,
+    /// undo groups that are undone
+    undos: BTreeSet<usize>,
+    /// undo groups that are no longer live and should be gc'ed
+    gc_undos: BTreeSet<usize>,
 
     this_edit_type: EditType,
     last_edit_type: EditType,
 
-    styles: Scopes,
-    doc_ctx: DocumentCtx,
-    config: BufferConfig,
     revs_in_flight: usize,
 
     /// Used only on Fuchsia for syncing
     #[allow(dead_code)]
     sync_store: Option<SyncStore>,
     #[allow(dead_code)]
     last_synced_rev: RevId,
-}
 
-#[derive(PartialEq, Eq, Clone, Copy)]
-enum EditType {
-    Other,
-    InsertChars,
-    Delete,
-    Undo,
-    Redo,
-    Transpose,
-}
-
-impl EditType {
-    pub fn json_string(&self) -> &'static str {
-        match *self {
-            EditType::InsertChars => "insert",
-            EditType::Delete => "delete",
-            EditType::Undo => "undo",
-            EditType::Redo => "redo",
-            EditType::Transpose => "transpose",
-            _ => "other",
-        }
-    }
+    syntax: SyntaxDefinition,
+    styles: Scopes,
+    config: BufferConfig,
 }
 
 impl Editor {
     /// Creates a new `Editor` with a new empty buffer.
-    pub fn new(doc_ctx: DocumentCtx, config: BufferConfig,
-               buffer_id: BufferIdentifier,
-               initial_view_id: ViewIdentifier) -> Editor {
-        Self::with_text(doc_ctx, config, buffer_id,
-                        initial_view_id, "".to_owned())
+    pub fn new(config: BufferConfig) -> Editor {
+        Self::with_text("".to_owned(), config)
```

Comment:
Is the `to_owned` necessary here, looks like the method has `Into` already. FWIW, I _really_ like that pattern and am in favor of using it more systematically (and using `.into()` when we don't want to make the methods generic, to signal that it's a straightforward conversion).

Reply:
roger. One thing I noticed during this is that rope is always constructed from a string slice; thinking of the big-file case, wouldn't it make sense to have a constructor that took an owned string, and just broke it into chunks without allocation? (I haven't looked at how amenable the builder is to this, just wanted to mention it while on my mind)

Reply:
So that's a deep line of inquiry. In the existing architecture, each leaf in the rope owns its chunk of string. I considered having a Cow, where the chunk would either be owned or a reference, but it increases complexity (especially managing the lifetime of the string it's referencing into), without clear benefits. The tradeoffs would be very different if we could memory-map a read-only snapshot, but for the most part this is not a primitive we have access to.

One small thing that could be done without refactor is to make the rope builder take `Into<Cow>`, which would let it avoid allocating a new owned string if the string is already suitable as a leaf.